snuba/subscriptions/worker.py

```diff
@@ -0,0 +1,49 @@
+from __future__ import annotations
+
+from typing import NamedTuple, Mapping, MutableSequence, Optional, Sequence
+
+from snuba.subscriptions.consumer import Tick
+from snuba.subscriptions.data import Subscription
+from snuba.subscriptions.executor import SubscriptionExecutor
+from snuba.subscriptions.scheduler import ScheduledTask, Scheduler
+from snuba.utils.streams.batching import AbstractBatchWorker
+from snuba.utils.streams.producer import Producer
+from snuba.utils.streams.types import Message, Topic
+from snuba.web.query import ClickhouseQueryResult
+
+
+class SubscriptionResult(NamedTuple):
+    task: ScheduledTask[Subscription]
+    result: ClickhouseQueryResult
+
+
+class SubscriptionWorker(AbstractBatchWorker[Tick, Sequence[SubscriptionResult]]):
```

Comment:
A bit of a high level question. What is the reason for making this a batching consumer ? Do we actually have a performance  gain in executing the queries in batches ? Or is this done in order to avoid committing every single offset ?

Reply:
Are we ok potentially producing twice the results for the same task on the same time frame in case this crashes and the batch is re-executed ?

Reply:
> What is the reason for making this a batching consumer ?

Mostly just that this maintains consistency with the other consumers. My thinking is not so much that there is a strong argument for making it a batching consumer, but more so that there is not really a strong argument for _not_ making it a batching consumer. There are likely some longer term improvements that can be made here as well that will be useful to other consumers (e.g. asynchronous commit, supporting concurrent message processing, etc.) as well.

> Do we actually have a performance gain in executing the queries in batches ?

Not at the moment. It is possible that coalescing queries into batches may be helpful.

> Or is this done in order to avoid committing every single offset ?

Not so much an "or", but yes, this is helpful to avoid having to reimplementing this behavior and configuration.

> Are we ok potentially producing twice the results for the same task on the same time frame in case this crashes and the batch is re-executed ?

I think we are less ok with not producing a result at all for a scheduled query, so we don't really have another option.

One drawback of the current implementation of the batching consumer is it doesn't allow a partial batch commit where the messages at the head of the batch that executed successfully can be committed (up to the first failure in each partition) which can lead to messages being processed (and in this case published) unnecessarily. That is one of those things that I would like to fix that would be broadly beneficial to other consumers as well.